batched/system.go

```diff
@@ -0,0 +1,388 @@
+package batched
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"strconv"
+	"sync"
+	"time"
+
+	"github.com/ipfs/go-cid"
+	"github.com/ipfs/go-datastore"
+	provider "github.com/ipfs/go-ipfs-provider"
+	"github.com/ipfs/go-ipfs-provider/queue"
+	"github.com/ipfs/go-ipfs-provider/simple"
+	logging "github.com/ipfs/go-log"
+	"github.com/ipfs/go-verifcid"
+	"github.com/multiformats/go-multihash"
+)
+
+var log = logging.Logger("provider.batched")
+
+type BatchProvidingSystem struct {
+	ctx     context.Context
+	close   context.CancelFunc
+	closewg sync.WaitGroup
+
+	reprovideInterval time.Duration
+	rsys              provideMany
+	keyProvider       simple.KeyChanFunc
+
+	q  *queue.Queue
+	ds datastore.Batching
+
+	reprovideCh chan cid.Cid
+
+	totalProvides, lastReprovideBatchSize     int
+	avgProvideDuration, lastReprovideDuration time.Duration
+}
+
+var _ provider.System = (*BatchProvidingSystem)(nil)
+
+type provideMany interface {
+	ProvideMany(ctx context.Context, keys []multihash.Multihash) error
+	Ready() bool
+}
+
+// Option defines the functional option type that can be used to configure
+// BatchProvidingSystem instances
+type Option func(system *BatchProvidingSystem) error
+
+var lastReprovideKey = datastore.NewKey("/provider/reprovide/lastreprovide")
+
+func New(provider provideMany, q *queue.Queue, opts ...Option) (*BatchProvidingSystem, error) {
+	s := &BatchProvidingSystem{
+		reprovideInterval: time.Hour * 24,
+		rsys:              provider,
+		keyProvider:       nil,
+		q:                 q,
+		ds:                datastore.NewMapDatastore(),
+		reprovideCh:       make(chan cid.Cid),
+	}
+
+	for _, o := range opts {
+		if err := o(s); err != nil {
+			return nil, err
+		}
+	}
+
+	if s.keyProvider == nil {
+		s.keyProvider = func(ctx context.Context) (<-chan cid.Cid, error) {
+			ch := make(chan cid.Cid)
+			close(ch)
+			return ch, nil
+		}
+	}
```

Comment:
We could error or require the keyProvider to exist in the constructor, but this seems fine. Either way we'd like to change the model here to use something other than the `keyProvider` function to specify what we are providing.

The changes here could be substantial so if there's a preference for make the keyProvider mandatory I'm fine with that too.

Reply:
This seems fine.